point-of-care.md

```diff
@@ -65,5 +65,34 @@ requirements a bit, the code may look like this:
 If we do this, we need a font where the 0 and the O are really different, or
 maybe underline the digits.
 
+### Another alternative
 
+We can generate 16 random bytes and use this as random authentication code.
 
+We can format this output as 6 numbers of 6 digits in length.
+
+This gives us an field of 10^36 which translates to an entropy of 
+ln(10)/ln(2) which is about 119 bits which puts us in the realm of the
+complexity of 3DES.
```

Comment:
IMO this is an overkill of security. When you make the tokens valid for a limited amount of time (which is not the case with 3DES) you can use codes with less digits. Typing in 36 digits is a hell of a job. Take some elderly person in mind they will drown in it.

Reply:
> When you make the tokens valid for a limited amount of time (which is not the case with 3DES) you can use codes with less digits. Typing in 36 digits is a hell of a job. Take some elderly person in mind they will drown in it.

This is a fair point....

But I don't think this is overkill, here's why:

First of all, because the whole country is looking at us for security.

Second consider the implications of what a malicious actor could do if they succeed in registering a malicious app. You could go "war-driving" and/or "war-walking" and order many people into quarantine.

If a person who is really infected with COVID-19 starts "war-walking" or "war-driving", I want all those people to self-quarantine.

But I also want the chances of someone guessing a correct activation code in the 30 minutes in which it is valid, to be so astronomically small that it can be considered near-impossible.

Reply:
> Second consider the implications of what a malicious actor could do if they succeed in registering a malicious app. You could go "war-driving" and/or "war-walking" and order many people into quarantine.
> 
> If a person who is really infected with COVID-19 starts "war-walking" or "war-driving", I want all those people to self-quarantine.

I am not seeing how this relates to the key-length. Can you explain more?

Reply:
> I am not seeing how this relates to the key-length. Can you explain more?

Simple:
6 digits gives us 1 million options.
log(2; 1000000 ^ 1) = about 20 bits of entropy.
log(2; 1000000 ^ 2) = about 40 bits of entropy.
log(2; 1000000 ^ 3) = about 60 bits of entropy.
log(2; 1000000 ^ 4) = about 80 bits of entropy.
log(2; 1000000 ^ 5) = about 100 bits of entropy.
log(2; 1000000 ^ 6) = about 120 bits of entropy.

The entire IPv6 address space is 2^128 addresses, if we get about 120 bits security and use a blocking policy of at most 10 wrong attempts per ip address, an attacker needs to send spoofed requests from 1 / (10*2^128)/(2^120) = 1/2560-th of the addresses within that space in an interval of 30 minutes.

If I drop to 5 groups of 6 digits the calculation becomes (1/(10*2^128)/(2^100)) = (1/2684354560), which is a subset of the address space that is small enough that I consider it a risk.

Also a 6x6 matrix of numbers just looks nice.

If we can have security near the realms of 3DES with this, I think we owe it to the people to give them that.

I also want the numbers to be "so far down in the dirt" to ensure that privacy-activists look at our back end implementation and can reasonably be said to "not have a serious argument".

Reply:
Another reason for choosing for 6 codes of 6 numbers, is that I want give a user the ability to "think about whether or not they want to enter those numbers" for a while.

That amount of time is 30 minutes right now, but it is a piece of cake to increase it to (for example) 7 days. With a code of 36 digits, the threat does not increase as quickly as it would with less than 36 digits.

Reply:
Something is going wrong here. I quoted your text about war-driving and I see it quoted in GitHub but in the mail another peace of text is quoted... weird.

I understand the reasoning behind why more digits are more secure (although I still think it's overkill and ruins the user experience). What I don't understand is the relation between key length and your story about war-driving. I understand that war-driving is a potential risk but not why a longer key is better here.

Reply:
> I understand the reasoning behind why more digits are more secure (although I still think it's overkill and ruins the user experience). What I don't understand is the relation between key length and your story about war-driving. I understand that war-driving is a potential risk but not why a longer key is better here.

Consider the following situation:

1. A healthcare worker creates a request_token for a couple of DayNumbers and dtks.
2. An attacker mounts a huge attack in which it sends us all possible possible tokens and a set of fake DayNumbers and dtks.
3. Our back end will accept the attacker's DayNumbers and dtks as real.
4. The user cannot transmit it's DayNumbers and dtks and is presented with an error, or the user does not submit it's DayNumbers and dtks at all.

The attacker now has a set of DayNumbers and dtks which are marked as "tested positive for COVID-19". With these keys in hand, the attacker can now deliberately trigger the "quarantine alert" on the devices of others within his vicinity.

And worst of all: We can only detect an attack like this when a user calls the healthcare worker whom has given the token to the user. This means that we cannot detect attacks that use tokens from users whom decide to not enter those numbers into the app.

Because a "You could be infected, go into quarantine-alert" has a huge impact on the lives of those who receive it, I want very strong security of a level comparable with that of 3DES.

In another scenario, we can have a malicious healthcare worker which deliberately marks a set of fake keys as positive.

This could be the end of the career of the healthcare worker in question! Especially if our government decides to make this app mandatory and/or attaches penalties to abuse, this can become especially problematic.

I want the chances of an attack as described above to be so incredibly low that I can confidently say that this is so improbable that I can confidently state that this attack cannot happen.

We can also instruct the healthcare workers to only give out codes to patients which are determined to enter them into the app, but people are fallible and people can lie as well.

Also: I expect that at most 20k users will have to enter those 36 numbers on any given day. If they can use a smartphone, they can probably enter a sequence which is shorter than 4 phone-numbers as well.

Note that I have not even started how this attack is similar to DNS-cache poisoning attacks that have been seen in the wild (I've attended talks by speakers that reported seeing these types of attacks on their own networks at a security-conference). And then there also is the issue about how being too lax with security could undermine the trust in this system as well.